datafusion/core/src/physical_plan/aggregates/row_hash.rs

```diff
@@ -761,4 +784,67 @@ impl GroupedHashAggregateStream {
         }
         Ok(Some(RecordBatch::try_new(self.schema.clone(), output)?))
     }
+
+    fn update_one_accumulator_with_native_value<T1>(
+        &mut self,
+        groups_with_rows: &[usize],
+        agg_input_array1: &T1,
+        acc_idx1: usize,
+        filter_bool_array: &[Option<&BooleanArray>],
+    ) -> Result<()>
+    where
+        T1: ArrowArrayReader,
+    {
+        let accumulator1 = &self.row_accumulators[acc_idx1];
+        let filter_array1 = &filter_bool_array[acc_idx1];
+        for group_idx in groups_with_rows {
+            let group_state = &mut self.aggr_state.group_states[*group_idx];
+            let mut state_accessor =
+                RowAccessor::new_from_layout(self.row_aggr_layout.clone());
+            state_accessor.point_to(0, group_state.aggregation_buffer.as_mut_slice());
+            for idx in &group_state.indices {
+                let value = col_to_value(agg_input_array1, filter_array1, *idx as usize);
+                accumulator1.update_value::<T1::Item>(value, &mut state_accessor);
+            }
+            // clear the group indices in this group
+            group_state.indices.clear();
+        }
+
+        Ok(())
+    }
+
+    fn update_two_accumulator2_with_native_value<T1, T2>(
```

Comment:
What is the reason to special case 2 accumulators?

Reply:
Yes, I think we have to handle the 2 accumulators cases specially,  to avoid the dynamic type dispatching in the inner loops and leverage generics to do the static type dispatching,  and also to achieve the best performance.

The current update flow is:

```
outmost loop :  for each `group idx` in groups_with_rows 
     construct the RowAccessor
     inner loop:s  for each accumuator
         update accumulator stats
```
The accumuators loop **must be** in the inner loop, so that we can benefit from the row layout and 
different accumulator stats are updated to the single row(the same `RowAccessor`).

And what we want to achieve is to do the array type cast out side the outmost loop

```
for each agg input arrays,  type cast here
outmost loop :  for each `group idx` in groups_with_rows
     construct the RowAccessor
     inner loop:s  for each accumuator
          call  generic method to update the accumuator stats
```

Because the agg input types for each accumulator is different, the generic method can not be called in the inner loops, so that I have to handle the 2 accumulators cases specially.   So the update flow becomes

1 Accumuator version:

```
agg input array1,  type cast to T1
outmost loop :  for each `group idx` in groups_with_rows 
     construct the RowAccessor
     generic method <T1>:  update accumuator stats
```

2 Accumuators version:

```
agg input array1,  type cast to T1
agg input array2,  type cast to T2
outmost loop :  for each `group idx` in groups_with_rows
     construct the RowAccessor
     generic method <T1>:  update accumuator1 stats
     generic method <T2>:  update accumuator2 stats
```

3 Accumuators version:

```
agg input array1,  type cast to T1
agg input array2,  type cast to T2
agg input array3,  type cast to T3
outmost loop :  for each `group idx` in groups_with_rows
     construct the RowAccessor
     generic method <T1>:  update accumuator1 stats
     generic method <T2>:  update accumuator2 stats
     generic method <T3>:  update accumuator3 stats

```

Hope this explains. I'm not sure whether there are other good ways to simply the code.

Reply:
> After thinking about this a bit -- I wonder if we can somehow get rid of the need to typecast during the aggregate at all -- it seems to me the code would be much simpler if the aggregator didn't have to cast its input, but instead the inputs were cast as needed.

I'm afraid the `typecast` can not be avoid.  This is because the arrow `Array` Trait itself does not provide any `read` functions to  read its value at a given `index`.  The trait must be downcast to the Struct(`PrimitiveArray<T:ArrowPrimitiveType>` or  `BooleanArray`) first.

I think this is related to `Trait Objects` restrictions.  If we want to add `generate` methods(like read value at given index) method or add `GAT` to a trait, that trait can not be used as a `Trait Objects` any more.
This is also why in this PR I had added a new trait. This new trait and its implementors bridge the type system between
the arrow `ArrowPrimitiveType` and the row accumulator internal state types.

```rust
pub trait ArrowArrayReader: Array {
    type Item: RowAccumulatorNativeType;

    /// Returns the element at index `i`
    /// # Panics
    /// Panics if the value is outside the bounds of the array
    fn value_at(&self, index: usize) -> Self::Item;

    /// Returns the element at index `i`
    /// # Safety
    /// Caller is responsible for ensuring that the index is within the bounds of the array
    unsafe fn value_at_unchecked(&self, index: usize) -> Self::Item;
}

impl<'a> ArrowArrayReader for &'a BooleanArray {
    type Item = bool;

    #[inline]
    fn value_at(&self, index: usize) -> Self::Item {
        BooleanArray::value(self, index)
    }

    #[inline]
    unsafe fn value_at_unchecked(&self, index: usize) -> Self::Item {
        BooleanArray::value_unchecked(self, index)
    }
}

impl<'a, T: ArrowPrimitiveType> ArrowArrayReader for &'a PrimitiveArray<T>
where
    <T as ArrowPrimitiveType>::Native: RowAccumulatorNativeType,
{
    type Item = T::Native;

    #[inline]
    fn value_at(&self, index: usize) -> Self::Item {
        PrimitiveArray::value(self, index)
    }

    #[inline]
    unsafe fn value_at_unchecked(&self, index: usize) -> Self::Item {
        PrimitiveArray::value_unchecked(self, index)
    }
}

```

Reply:
Actually in. the current arrow-rs compute kernels, the `typecast` is everywhere

https://github.com/apache/arrow-rs/blob/b163b19d213c57170789f32a2011cbadf9ab4120/arrow-ord/src/comparison.rs#L1598-L1748

Reply:
Sorry -- I mis understood -- I thought you meant like when aggregating SUM(Int16) that the argument needed to be cast to `Int32` or something as the type of the aggregator was different than the type of the inderlying value in the accumulator

I agree the downcasting is needed to figure out the correct function to call -- though the more I think about it the more I think it can / should be done during setup time and not on every batch (as the types don't change during the query execution)